File: tests/reboot/documentation/auth_examples.py
from google.protobuf.message import Message
from rbt.v1alpha1 import errors_pb2
from reboot.aio.applications import Application
from reboot.aio.auth import Auth
from reboot.aio.auth.authorizers import allow, allow_if
from reboot.aio.auth.token_verifiers import TokenVerifier, VerifyTokenResult
from reboot.aio.contexts import ReaderContext
from reboot.aio.external import ExternalContext
from tests.reboot.documentation.bank_rbt import Account
from typing import Optional


async def validate_user(user_id: str) -> bool:
    return user_id != ""


async def is_valid_user(
    *,
    context: ReaderContext,
    state: Optional[Message],
    request: Optional[Message],
    **kwargs,
):
    if context.auth is None or context.auth.user_id is None:
        return errors_pb2.Unauthenticated()

    if await validate_user(context.auth.user_id):
        return errors_pb2.Ok()

    return errors_pb2.PermissionDenied()


async def is_admin(
    *,
    context: ReaderContext,
    state: Optional[Message],
    request: Optional[Message],
    **kwargs,
):
    if context.auth is not None and context.auth.user_id == "admin":
        return errors_pb2.Ok()

    return errors_pb2.PermissionDenied()


async def is_account_owner(
    *,
    context: ReaderContext,
    state: Optional[Message],
    request: Optional[Message],
    **kwargs,
):
    if context.auth is not None and context.auth.user_id == context.state_id:
        return errors_pb2.Ok()

    return errors_pb2.PermissionDenied()


class AccountServicer(Account.Servicer):

    def authorizer(self):
        return Account.Authorizer(
            balance=allow_if(any=[is_admin, is_account_owner]),
            deposit=allow(),
            withdraw=allow_if(all=[is_account_owner]),
        )


class AdminOnlyAccountServicer(Account.Servicer):

    def authorizer(self):
        return allow_if(all=[is_admin])


class MyTokenVerifier(TokenVerifier):

    async def verify_token(
        self,
        context: ReaderContext,
        token: Optional[str],
    ) -> VerifyTokenResult:
        if token is None:
            return None
        return Auth(user_id=token)


async def main():
    application = Application(
        servicers=[AccountServicer],
        token_verifier=MyTokenVerifier(),
    )
    await application.run()


def make_context(token: str) -> ExternalContext:
    context = ExternalContext(
        name="Example",
        url="http://localhost:9991",
        bearer_token=token,
    )
    return context

File: tests/reboot/documentation/auth_examples.ts
import { Message } from "@bufbuild/protobuf";
import {
  allow,
  allowIf,
  Application,
  Auth,
  ExternalContext,
  ReaderContext,
  TokenVerifier,
  WriterContext,
} from "@reboot-dev/reboot";
import { errors_pb } from "@reboot-dev/reboot-api";
import { Account, Bank } from "./bank_rbt.js";

const SINGLETON_BANK_ID = "singleton-bank";
const VALID_JWT = "a-valid-json-web-token";

async function validateUser(userId: string): Promise<boolean> {
  return userId !== "";
}

async function isValidUser({
  context,
}: {
  context: ReaderContext;
  state?: Message;
  request?: Message;
}) {
  if (!context.auth) {
    return new errors_pb.Unauthenticated();
  }

  if (await validateUser(context.auth.userId)) {
    return new errors_pb.Ok();
  }

  return new errors_pb.PermissionDenied();
}

async function isAdmin({ context }: { context: ReaderContext }) {
  if (context.auth?.userId === "admin") {
    return new errors_pb.Ok();
  }

  return new errors_pb.PermissionDenied();
}

async function isAccountOwner({ context }: { context: ReaderContext }) {
  if (context.auth?.userId === context.stateId) {
    return new errors_pb.Ok();
  }

  return new errors_pb.PermissionDenied();
}

class AccountServicerBase extends Account.Servicer {
  async open(
    context: WriterContext,
    request: Account.OpenRequest
  ): Promise<Account.PartialOpenResponse> {
    this.state.customerName = request.customerName;
    return {};
  }

  async getId(
    context: ReaderContext,
    request: Account.GetIdRequest
  ): Promise<Account.PartialGetIdResponse> {
    return { id: context.stateId };
  }

  async balance(
    context: ReaderContext,
    request: Account.BalanceRequest
  ): Promise<Account.PartialBalanceResponse> {
    return { balance: this.state.balance };
  }

  async deposit(
    context: WriterContext,
    request: Account.DepositRequest
  ): Promise<Account.PartialDepositResponse> {
    this.state.balance += request.amount;
    return { updatedBalance: this.state.balance };
  }

  async withdraw(
    context: WriterContext,
    request: Account.WithdrawRequest
  ): Promise<Account.PartialWithdrawResponse> {
    this.state.balance -= request.amount;
    return { updatedBalance: this.state.balance };
  }
}

export class AccountServicer extends AccountServicerBase {
  authorizer() {
    return new Account.Authorizer({
      balance: allowIf({ any: [isAdmin, isAccountOwner] }),
      deposit: allow(),
      withdraw: allowIf({ all: [isAccountOwner] }),
    });
  }
}

export class AdminOnlyAccountServicer extends AccountServicerBase {
  authorizer() {
    return allowIf({ all: [isAdmin] });
  }
}

export class MyTokenVerifier extends TokenVerifier {
  async verifyToken(
    context: ReaderContext,
    token?: string
  ): Promise<Auth | null> {
    if (token === undefined) {
      return null;
    }
    return new Auth({ userId: token });
  }
}

export function main() {
  new Application({
    servicers: [AccountServicer],
    tokenVerifier: new MyTokenVerifier(),
  }).run();
}

export function makeContext(token: string): ExternalContext {
  const context = new ExternalContext({
    name: "Example",
    url: "http://localhost:9991",
    bearerToken: token,
  });
  return context;
}

export async function createBank(context: ExternalContext) {
  const [bank] = await Bank.create(
    context,
    SINGLETON_BANK_ID,
    {},
    { bearerToken: VALID_JWT }
  );
  return bank;
}

export function refAccount(request: Bank.TransferRequest, bearerToken: string) {
  const fromAccount = Account.ref(request.fromAccountId, { bearerToken });
  return fromAccount;
}
